---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 1/12/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
``` {r setup}
 knitr::opts_knit$set(root.dir = normalizePath("/Users/danasouter/Desktop/qbio490/qbio_490_dana/analysis_data"))
```

Load all necessary packages and instantiate the rna_se object
```{r}
#packages
#BiocManager::install("DESeq2")
#BiocManager::install("EnhancedVolcano")

library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)
```

```{r}
#GDCDownload()
```


```{r}
#instantiate object
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable)

For rna_clinical:
```{r}
#creation:
#rna_clinical <- colData(rna_se)
#colnames(rna_clinical) <- c("Sample_ID", "Gender", "Age", "Tumor_Status")

rna_clinical <- rna_se@colData

#removal of "nested lists":
#rna_clinical <- data.frame(lapply(rna_clinical, unlist))
rna_clinical <- as.data.frame(rna_clinical)

colnames(rna_clinical)
rna_clinical <- rna_clinical[ , 25]


```

For rna_genes:
```{r}
#creation:
#rna_genes <- assay(rna_se)
#rownames(rna_genes) <- rownames(rna_se)

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)


#changing rownames
#rna_genes <- data.frame(Sample_ID = rownames(rna_genes), rna_genes)

row.names(rna_genes) <- rna_genes$gene_id

```

For rna_counts:
```{r}
#creation:
#rna_counts <- assay(rna_se)
#rownames(rna_counts) <- rownames(rna_se)
#colnames(rna_counts) <- colnames(rna_se)

#changin row names and colnames
#rna_counts <- data.frame(Sample_ID = rownames(rna_counts), rna_counts)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

row.names(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rna_clinical$barcode

```


*2*
Determine and define your experimental design for the differential expression analysis, comparing patients samples by gender. Control for covariance of age_category OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?) Filter out genes with a total expression across all patients less than 10.
```{r}
sum(is.na(rna_clinical$age_category))
rna_clinical$vital_status[244]

NR_vital_mask <- ifelse(rna_clinical$vital_status == "Not Repoted", F, T)
clean_clinical <- rna_clinical[NR_vital_mask]
clean_clinical$vital_status <- factor(clean_clinical$vital_status)

NA_age_mask <- !is.na(clean_clinical$age_at_index)
clean_clinical <- clean_clinical[NA_age_mask]
clean_clinical$age_category <- ifelse(clean_clinical$age_at_index >= 58, "old", "young")
clean_clinical$age_category <- factor(clean_clinical$age_category)

sum(is.na(clean_clinical$gender))
clean_clinical$gender <- factor(clean_clinical$gender)


clean_counts <- rna_counts[,NR_vital_mask]
clean_counts <- clean_counts[,NA_age_mask]

less_10_mask <- rowSums(rna_counts) >= 10
clean_counts <- clean_counts[less_10_mask, ]
clean_genes <- rna_genes[less_10_mask, ]

#if ("age_category" %in% colnames(rna_clinical)) {
 # if (length(unique(rna_clinical$age_category)) == 0) {
  #  print("age_category column exists but contains no values. Please #check the data source or loading process.")
 # } else {
 #   rna_clinical$age_category <- as.factor(rna_clinical$age_category)
 # }
#} else {
#  print("age_category column does not exist in rna_clinical dataframe.")
#}

#if ("vital_status" %in% colnames(rna_clinical)) {
#  if (length(unique(rna_clinical$vital_status)) == 0) {
 #   print("vital_status column exists but contains no values. Please #check the data source or loading process.")
#  } else {
#    rna_clinical$vital_status <- as.factor(rna_clinical$vital_status)
#  }
#} else {
#  print("vital_status column does not exist in rna_clinical dataframe.")
#}

#total_expression <- rowSums(rna_counts[, -1])  
#rna_counts_filtered <- rna_counts[rowSums(rna_counts[, -1]) >= 10, ] 
#design <- ~ gender + age_category  
#head(rna_counts_filtered)

```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
#ncol(rna_counts_filtered)
#nrow(rna_counts_filtered)

#nrow(rna_clinical)
#ncol(rna_clinical)

dds <- DESeqDataSetFromMatrix(countData = clean_counts,
                              colData = clean_clinical,
                              design =~age_category + gender + vital_status)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("vital_status", "Dead", "Alive"))

results <- data.frame(results)
```

Prepare results dataframe for EnhancedVolcano plotting. 
```{r}
#adding information columns and -log10(padj) column
results$gene_name <- rna_genes$gene_name
results$"-log10(padj)" <- -log10(results$padj) 

row.names(results) <- rna_genes$gene_id

results
```


```{r}
EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = 'pvalue',
                title = 'Vital Status: Alive vs Dead',
                pointSize = 1.0,
                labSize = 5.0)
```


# Explain what genes from each part of the Volcano Plot mean in terms of their significance and up/down regulation. 
top-right genes: upregulated in the comparison group and high significance
bottom-right genes: upregulated and pretty significant. 
top-left genes: significantly downregulated in the comparison group and are significant. 
bottom-left genes: downregulated and slightly significants
top-middle genes: highly significant but not significantly upregulated or downregulated.
bottom-middle genes: not significant in pretty much any way.

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
